File: package/go/src/run_length_encoding.go
package src

// RunLengthEncodingResult represents a character and its count
type RunLengthEncodingResult struct {
	Char  string
	Count int
}

// RunLengthEncoding performs run-length encoding on a string
func RunLengthEncoding(s string) []RunLengthEncodingResult {
	if len(s) == 0 {
		return []RunLengthEncodingResult{}
	}

	var result []RunLengthEncodingResult
	currentChar := string(s[0])
	count := 1

	for i := 1; i < len(s); i++ {
		char := string(s[i])
		if char == currentChar {
			count++
		} else {
			// Character changed, add current group to result
			result = append(result, RunLengthEncodingResult{
				Char:  currentChar,
				Count: count,
			})
			currentChar = char
			count = 1
		}
	}

	// Add the last group
	result = append(result, RunLengthEncodingResult{
		Char:  currentChar,
		Count: count,
	})

	return result
}

File: package/typescript/src/runLengthEncoding.ts
/**
 * 文字列 s をランレングス符号化する関数
 *
 * @param s - 符号化する文字列（英小文字のみ、1文字以上）
 * @returns 連続する文字とその回数のタプルからなるリスト
 */
export function runLengthEncoding(s: string): Array<[string, number]> {
  if (!s) {
    return [];
  }

  const result: Array<[string, number]> = [];
  let currentChar = s[0];
  let count = 1;

  for (let i = 1; i < s.length; i++) {
    if (s[i] === currentChar) {
      count++;
    } else {
      result.push([currentChar, count]);
      currentChar = s[i];
      count = 1;
    }
  }

  // Add the last group
  result.push([currentChar, count]);

  return result;
}
